Add fix to override afterRemote comfirm method

PradnyaBaviskar opened the pull request:
to / @bajtos -
In reference to https://github.com/strongloop/loopback/issues/416 , 

I've made changes so that redirect URL is an optional param. And in case redirect URL is specified then the response status is 302.

For this to work correctly, the strong remoting fix - https://github.com/strongloop/strong-remoting/pull/183, will have to be taken up first.

Can you please review?

bajtos commented on common/models/user.js:
```diff
@@ -591,7 +591,11 @@ module.exports = function(User) {
+          next();
         } else {
           next(new Error('transport unsupported'));
         }
```
While you are changing this part, please rework the implementation as follows:

 1. Use `ctx.args.redirectUrl` to obtain the parameter value, don't use `ctx.req` directly.
 2. Don't throw `Error("transport unsupported")` when `redirectUrl` is not set

bajtos commented:
@slnode test please

bajtos commented on test/user.test.js:
```diff
@@ -858,6 +858,31 @@ describe('User', function() {
+              '&redirect=' + encodeURIComponent(options.redirect))
+            .expect(302);
+
+          done();
```
You have to pass `done` as a callback of `expect` and/or `end`.

The current implementation marks the test as done before the HTTP request is finished.

The same applies to the next test below.

```js
.expect(302)
.end(done);
```

PradnyaBaviskar commented:
@bajtos,

I've modified the code, so as to use ```ctx.args.redirectUrl```, rather than using ```ctx.req```.
Also modified the test cases as suggested above.

Please review.

bajtos commented on common/models/user.js:
```diff
@@ -580,21 +580,11 @@ module.exports = function(User) {
+          ctx.res.location(ctx.args.redirect);
+          ctx.res.status(302);
         }
+        next();
```
`ctx.res` may be undefined when non-REST transport like WebSockets is used.

Here is a mockup of what I'd like to see here: 

```js
if (ctx.args.redirect !== undefined) {
  if (!ctx.res) {
    return next(new Error('The transport does not support HTTP redirects.'));
  }
  // redirect via ctx.res.location & status
}
next();
```

PradnyaBaviskar commented:
Made changes to handle undefined ```ctx.res``` for non-REST transports.

@bajtos, can you review?

bajtos commented on common/models/user.js:
```diff
@@ -580,21 +580,14 @@ module.exports = function(User) {
-        } else {
-          next(new Error('transport unsupported'));
+        if (ctx.args.redirect !== undefined) {
+          if(!ctx.res) {
```
style: missing space after "if" keyword (see `npm test`)

bajtos commented:
Please update the strong-remoting version in package.json to `^2.13.2`.

bajtos commented on test/user.test.js:
```diff
@@ -858,6 +858,38 @@ describe('User', function() {
+                return done(err);
+              }
+              done();
+            });
```
Since you are not making any extra checks here, the code can be simplified to

```js
  .expect(204)
  .end(done);
```

bajtos commented on test/user.test.js:
```diff
@@ -858,6 +858,38 @@ describe('User', function() {
+                return done(err);
+              }
+              done();
+            });
```
Same comment here, please simplify to 

```js
  .expect(Location, ...)
  .end(done);
```

bajtos commented:
I have improved the commit message and landed the patch via 3c43ecc. Thank you for the contribution!
